### DISCUSSION TITLE
Date strings with fractions of a second marked as invalid when formatting

### DISCUSSION THREAD

**Discussion by PontusMagnusson** [js-fns/date-fns#2596] [General] [answered]
I'm trying move away from using moment.js and as I'm trying to format some dates that are in my system, date-fns tells me that the dates are of an invalid format.

I have a function that I use to format dates which looks as follows:
```js
import { format } from "date-fns";

export function formatDateTime(date) {
  if (date) return format(date, "yyyy-MM-dd HH:mm:ss");
}
```

When trying to format this date string `2021-08-09T10:52:37.6532978` it throws an error and tells me the date is invalid, but moment.js handles it just fine.

It handles `2021-05-07T14:54:53`  perfectly fine, so I assume it has something to do with the second fractions at the end.

If I use `utcToZonedTime` from `date-fns-tz` on the failing string first, it manages to understand the date string.

Is there something I'm missing? Thanks.

Edit: 

```js
import { format } from "date-fns";

export function formatDateTime(date) {
  if (date) return format(new Date(date), "yyyy-MM-dd HH:mm:ss");
}
```
Creating a new Date object seems to work, but I'm not sure why just passing the string throws an error.

**Comment by fturmel** [comment, marked as answer]
`format` only accepts a Date instance or a number (ms since Unix epoch) as its first argument. If you start with a string representation of a date, you need to parse it into a Date object first (with the `parse` or `parseISO` functions for example).
https://date-fns.org/v2.23.0/docs/format

Careful using strings in the Date constructor, it is discouraged by MDN because the implementation varies with each runtime.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date

**Reply by PontusMagnusson** [reply]
Ah, thank you. I've misunderstood the documentation a bit. That the function sometimes accepts strings also confused me.

**Reply by fturmel** [reply]
`format` should not work with a date string in date-fns v2 for any scenarios, it will throw a RangeError.

In v1 however, `format` did accept and parse strings internally. This is one of the major changes between the two versions.
https://date-fns.org/v2.23.0/docs/Upgrade-Guide#string-arguments

So, maybe you referenced v1 docs or examples at some point? Or maybe you've been using v1 all along? If you have examples of v2 `format` working with a string date I'd love to see it.

**Reply by PontusMagnusson** [reply]
Yeah that was a mistake on my part. I noticed that some of the strings got parsed in another component, so the docs are of course correct.

Thanks for the help :)

**Reply by fturmel** [reply]
You're welcome, and thanks for confirming!
